modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -79,21 +79,333 @@ interface DomParser {
+          textNode = textNode.prev;
+          continue;
+        }
+      }
+
+      const sibling = textNode.prev;
+      textNode.remove();
+      textNode = sibling;
+    }
+  };
+
+  const cloneAndExcludeBlocks = (input: SchemaMap) => {
+    const output: SchemaMap = {};
+
+    for (const name in input) {
+      if (name !== 'li' && name !== 'p') {
+        output[name] = input[name];
+      }
+    }
+
+    return output;
+  };
+
+  const parser = SaxParser({
+    validate: settings.validate,
+    document: settings.document,
+    allow_html_data_urls: settings.allow_html_data_urls,
+    allow_svg_data_urls: settings.allow_svg_data_urls,
+    allow_script_urls: settings.allow_script_urls,
+    allow_conditional_comments: settings.allow_conditional_comments,
+    preserve_cdata: settings.preserve_cdata,
+
+    // Exclude P and LI from DOM parsing since it's treated better by the DOM parser
+    self_closing_elements: cloneAndExcludeBlocks(schema.getSelfClosingElements()),
+
+    cdata: (text) => {
+      node.append(new AstNode('#cdata', 4)).value = text;
+    },
+
+    text: (text, raw) => {
+      let textNode;
+
+      // Trim all redundant whitespace on non white space elements
+      if (!isInWhiteSpacePreservedElement) {
+        text = text.replace(allWhiteSpaceRegExp, ' ');
+
+        if (isLineBreakNode(node.lastChild, blockElements)) {
+          text = text.replace(startWhiteSpaceRegExp, '');
+        }
+      }
+
+      // Do we need to create the node
+      if (text.length !== 0) {
+        textNode = new AstNode('#text', 3);
+        textNode.raw = !!raw;
+        node.append(textNode).value = text;
+      }
+    },
+
+    comment: (text) => {
+      node.append(new AstNode('#comment', 8)).value = text;
+    },
+
+    pi: (name, text) => {
+      node.append(new AstNode(name, 7)).value = text;
+      removeWhitespaceBefore(node);
+    },
+
+    doctype: (text) => {
+      const newNode = node.append(new AstNode('#doctype', 10));
+      newNode.value = text;
+      removeWhitespaceBefore(node);
+    },
+
+    start: (name, attrs, empty) => {
+      const newNode = new AstNode(name, 1);
+      newNode.attributes = attrs;
+      newNode.shortEnded = empty;
+
+      node.append(newNode);
+
+      // Trim whitespace before block
+      if (blockElements[name]) {
+        removeWhitespaceBefore(newNode);
+      }
+
+      // Change current node if the element wasn't empty i.e not <br /> or <img />
+      if (!empty) {
+        node = newNode;
+      }
+
+      // Check if we are inside a whitespace preserved element
+      if (!isInWhiteSpacePreservedElement && whiteSpaceElements[name]) {
+        isInWhiteSpacePreservedElement = true;
+      }
+    },
+
+    end: (name) => {
+      let textNode, text, sibling, tempNode;
+
+      const elementRule = settings.validate && schema.getElementRule(node.name) || {} as SchemaElement;
```

Comment:
You're still doing schema validation here by the looks?

Reply:
This used to be this:
```typescript
const elementRule: Partial<SchemaElement> = validate ? schema.getElementRule(name) : {};
```
which would validate using the schema. What we have now, with the `schema.getElementRule(node.name) || {}` doesn't involve any validation — it just looks at the schema (with fallback) to inform the whitespace / empty-element removal logic

Reply:
```
const elementRule = settings.validate && schema.getElementRule(node.name) || {} as SchemaElement;
```
is the same as what you said the original was?
```
const elementRule: Partial<SchemaElement> = validate ? schema.getElementRule(name) : {};
```
as in the `validate` true case you're still going to end up looking up the schema and getting the element rule and in the false case you're going to use the empty object.

For this to be the same as a native parser it cannot use either `validate` or `schema.getElementRule`.

Reply:
They both look up the element rule, the difference is in what they do when that element rule can't be found / doesn't exist. Previously it would end up with `const elementRule = null` and then the node gets completely skipped. Now you'd end up with `const elementRule = null || {}`, which means the node still gets processed we just don't apply any specific logic to it. We do need to move towards it not referencing anything at all, which is what I'm working on now on a branch off of this, but it hasn't happened yet.

Reply:
Right I'd missed that because it's a bit hacky and not something we'd normally do. Anyways, I've looked at this more and the `elementRule` is only used for adding padding or removing empty elements. As we said on Friday that can't be in the builder section and needs to be moved out to a later section. So I can't approve this until we've done the separation properly sorry.

Reply:
Alternatively we need to change the scope of this so it's only meant to be a step 1 of separating the logic into the various components/steps that we're going to require moving forward.

Reply:
Discussed this more at stand up and we're moving to the iterative approach whereby this single change can't be released without other PRs.